### PR TITLE
Revert "Merge pull request #19583 from gnossen/revert_signal_handling"

### PR DESCRIPTION
**Pull request by gnossen**
This reverts commit 1e7ec75eff60ff74d0c192591a369af0308bca48, reversing
changes made to 6d62eb1b703617ff9165773b6d1e7d28ab84856d.

### CODE REVIEW

File: src/python/grpcio_tests/tests/unit/_signal_handling_test.py [hunk 1]
```diff
@@ -27,7 +27,18 @@
 from tests.unit import test_common
 from tests.unit import _signal_client
 
-_CLIENT_PATH = os.path.abspath(os.path.realpath(_signal_client.__file__))
+_CLIENT_PATH = None
+if sys.executable is not None:
+    _CLIENT_PATH = os.path.abspath(os.path.realpath(_signal_client.__file__))
+else:
+    # Note(rbellevi): For compatibility with internal testing.
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/unit/_signal_handling_test.py]: s/Note/NOTE

**Comment by gnossen** [comment]
Interop failure: https://github.com/grpc/grpc/issues/18616

File: src/python/grpcio_tests/tests/unit/_signal_handling_test.py [hunk 2]
```diff
@@ -27,7 +27,18 @@
...
+    if len(sys.argv) != 2:
+        raise RuntimeError("Must supply path to executable client.")
+    client_name = sys.argv[1].split("/")[-1]
+    del sys.argv[1]  # For compatibility with test runner.
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/unit/_signal_handling_test.py]: Just curious, what will happen if we don't delete the `argv[1]`?

File: src/python/grpcio_tests/tests/unit/_signal_handling_test.py [hunk 3]
```diff
@@ -104,6 +115,15 @@ def _read_stream(stream):
     return stream.read()
 
 
+def _start_client(args, stdout, stderr):
+    invocation = None
+    if sys.executable is not None:
+        invocation = (sys.executable, _CLIENT_PATH) + tuple(args)
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/unit/_signal_handling_test.py]: Interestingly, concatenation on tuples is faster than lists.

```
>>> import timeit
>>> a =[1,3,4,5]
>>> def t1():
...   return (7,) + tuple(a)
...
>>> timeit.timeit(t1)
0.21551179885864258
>>> def t2():
...   return [7] + a
...
>>> timeit.timeit(t2)
0.2819240093231201
```

File: src/python/grpcio_tests/tests/unit/_signal_handling_test.py (hunk 2 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests/unit/_signal_handling_test.py]: Our internal test runner will choke on the argument.

**Pull request merged**